editor/utils/block-list.js

```diff
@@ -21,22 +21,36 @@ const INNER_BLOCK_LIST_CACHE = {};
  * because otherwise the rendering of a nested BlockList will cause ancestor
  * blocks to re-mount, leading to an endless cycle of remounting inner blocks.
  *
- * @param {string}   uid                   Block UID to use as root UID of
- *                                         BlockList component.
- * @param {Function} renderBlockMenu       Render function for block menu of
- *                                         nested BlockList.
- * @param {boolean}  showContextualToolbar Whether contextual toolbar is to be
- *                                         used.
+ * @param {string}   uid                   	          Block UID to use as root UID of
+ *                                         	          BlockList component.
+ * @param {Function} renderBlockMenu       	          Render function for block menu of
+ *                                         	          nested BlockList.
+ * @param {boolean}  showContextualToolbar 	          Whether contextual toolbar is to be
+ *                                         	          used.
+ * @param {Function} updateNestedSettingsIfNeeded     Function that updates nesting settings
+ *                                                    if an update is necessary
+ * @param {Function} insertTemplateBlockIfNeeded      Function that inserts blocks in accordance
+ *                                                    with the template if innerBlocks are empty
  *
  * @return {Component} Pre-bound BlockList component
  */
-export function createInnerBlockList( uid, renderBlockMenu, showContextualToolbar ) {
+export function createInnerBlockList( uid, renderBlockMenu, showContextualToolbar, updateNestedSettingsIfNeeded, insertTemplateBlockIfNeeded ) {
 	if ( ! INNER_BLOCK_LIST_CACHE[ uid ] ) {
 		INNER_BLOCK_LIST_CACHE[ uid ] = [
 			// The component class:
 			class extends Component {
 				componentWillMount() {
 					INNER_BLOCK_LIST_CACHE[ uid ][ 1 ]++;
+					updateNestedSettingsIfNeeded( {
+						supportedBlocks: this.props.allowedBlockNames,
+					} );
+					insertTemplateBlockIfNeeded( this.props.template );
+				}
+
+				componentWillReceiveProps( nextProps ) {
+					updateNestedSettingsIfNeeded( {
```

Comment:
As far As I understand we have the `uid` at this point, so we can call the action directly in this component to avoid the need to add a new callback to this HoC?